### DISCUSSION TITLE
Add a form button that downloads an attachment to an autosubmit form

### DISCUSSION THREAD

**Discussion by spapas** [unpoly/unpoly#518] [Q&A] [answered]
Hello, I've got a form like this:
```html
<form class='form-inline' method='GET' action='' up-autosubmit 
up-watch-delay="150" up-history up-disable
up-target='.table-container'>
	{{ filter.form|crispy }}
	<input type='submit' name='show' class='btn btn-info' value='Show' />
	<button type='submit' name='download' class='btn btn-info'>
		Download
	</button>
</form>
```

The download button will return an xls file with a content-disposition=attachment. However when I click it it will return the xls file in the browser window dispalying garbage (the content disposition isn't respected?)

What I want is to somehow "disable" the autosubmit when that button is clicked i.e when the Download button is clicked I want it to behave like the form is not unpoly-enabled. I tried adding options like up-submit=false or up-follow=false or download to the button but nothging happened.

Thank you

**Comment by adam12** [comment]
Unpoly might be able to check for `content-disposition=attachment` in the response and then make a real request instead, but I am not sure we'd want to make that the default (the request itself might be super expensive).

Untested, but I'd use a listener on `up:form:submit` which cancels the request and makes a real request instead.

```js
up.on('up:form:submit', (event, form) => {
  if (event.submitButton.name === 'download') {
    event.preventDefault() // Prevent Unpoly from performing this submission
    form.submit() // Trigger manual form submission using browser
  }
})
```

EDIT: Thinking about this a bit more, I see two issues with my suggestion: a simple `form.submit()` will not be enough to indicate the button used to do the submission (maybe you can work around this with a manually created [`SubmitEvent`](https://developer.mozilla.org/en-US/docs/Web/API/SubmitEvent/SubmitEvent)).  The second issue is that `form.submit()` or a `SubmitEvent` might trigger this listener again, causing an infinite loop.

**Reply by spapas** [reply, marked as answer]
Hey @adam12 thank you for answering but as you already found out, `form.submit()` will trigger the form submission from the 1st and not the 2nd (download) button so it won't work :( 

I did some more research and found out this solution:

```
<input type='submit' name='show' class='btn btn-info' value='Εμφάνιση' onclick='this.form.target="_self";'  />
<input type='submit' name='download' class='btn btn-info' value='Λήψη' onclick='this.form.target="_blank";'  />
```

Please notice the difference between the buttons (target='_self' and '_blank'). The download button now tries to open in a new window resulting in the download attachment working properly!

**Reply by adam12** [reply]
Nice solution :) glad you were able to find a fix.

**Reply by licht1stein** [reply]
When using this solution, after a blank page opens and file downloads (that part works), other buttons also start showing their response in new window. 

Maybe there's a way around it?

**Reply by spapas** [reply]
@licht1stein doesn't it work if you add onclick='this.form.target="_self"; to the other buttons?

**Reply by licht1stein** [reply]
Nope, that's exactly what I did.

**Comment by triskweline** [comment]
A riff on [@adam12's solution](https://github.com/unpoly/unpoly/discussions/518#discussioncomment-6605346), using `up.network.loadPage()` instead of `form.submit()`:

```js
up.on('up:form:submit', (event, form) => {
  if (event.submitButton.name === 'download') {
    event.preventDefault() // Prevent Unpoly from handling this submission
    up.network.loadPage(event.renderOptions) // Submit form params with a full page load
  }
})
```

If the submit button has a `[name]` and `[value]` attribute, it will be included in the params sent to the server by `up.network.loadPage()`. This way the server can understand that a download is being requested. As an alternative, the event listener could also mutate `event.renderOptions.params`.

One limitation of `up.network.loadPage()` is that it can only send textual param values, so any file inputs will be excluded.
